Create missing params when reading from excel

trevorb1 opened the pull request:
Hello! 

# Overview
In this PR I implement logic in the `ReadExcel` class to create empty dataframes for parameters that do not have an associated excel sheet. This PR relates to issue #99, and [this comment](https://github.com/OSeMOSYS/otoole/issues/114#issuecomment-1358393751) in issue #114. 

# Problem Addressed 
I was reviewing issue #99 and encountered a very similar issue to the one described. When we convert from anything to an Excel file, only parameters that have non-default values are written. For example, in Simplicity the DiscountRate is given by a default value, therefore, no `DiscountRate` sheet will appear after running the command `otoole convert <> excel <> params.xlsx config.yaml`. 

In a different [comment](https://github.com/OSeMOSYS/otoole/issues/114#issuecomment-1358393751) (see issue #114) I discuss if this is actually the functionality we want. Or if we should be writing out empty sheets when the parameter does not have values to help guide users on how to input data. 

Nevertheless, with the current functionality, when reading in the Excel data it is not capturing parameters without dedicated sheets. For example, if no `DiscountRate` sheet exists in the excel file, then no `DiscountRate` key will be created in the otoole internal datastore. While this doesn't cause any otoole logic issues, no reference to the missing parameters will be written. This is problematic, as, for example, you will have an OSeMOSYS datafile without a `DiscountRate` parameter.   

# Logic
To fix this, I add a new function to the `ReadExcel` class called `_get_missing_params( ... )`. This function takes in the parameters read in from excel and compares them against all defined parameters in the `config.yaml` file. If a parameter is missing in the excel data, an empty dataframe is created and added to the internal datastore. Two tests are added to the `test_read_strategies.py` module to accompany this logic. 

# Other notes
Currently, this only checks parameters. A minor modification is needed if we want it to check for missing sets as well. However, I decided to hold off on adding that until we discussed if this is the correct path to go down to fix this issue :) 

Thanks!

willu47 reviewed (commented):
Hi @trevorb1 - this also looks good. However, couldn't all types of input formats have missing parameters?  In which case, could we bubble this function up to the class level so that missing parameters are created for all Read subclasses?

trevorb1 commented:
Hi @willu47! Good call, I think that makes way more sense! I have updated this PR to generalize the  `_get_missing_params( ... )` function (issue #99), and provide excel templates for missing parameters (issue #114). These two issues are grouped into one PR since they cover similar issues; however, if quite a few more changes needs to happen from this point I am happy to break them up. 

# Issue 99 Updates
In the commit "added function to get missing input values", I move the `_get_missing_params( ... )` function to the `ReadStrategy` class, rename it to `_get_missing_input_dataframes(...)`, and generalize it to work with any of `param`, `sets`, and `result` types. A test is added as well to the `TestReadStrategy` class. 

## Questions 
1. The function `_get_missing_input_dataframes(...)` will create empty, but correctly formatted, dataframes for any missing `param`, `set` or `result` that is listed in the config file but is not found in the internal datastore after reading in data. To keep this general, I have the user pass in an argument for what parameter to search for (`param`, `set` or `result`). 

My question is, I often have to loop over two of these items (see comment below) which seems a little clunky. Do you think there is a way to simplify this or change the logic at all? My first thought was to just accept a list of inputs to loop over, instead of a single string. However I just move the loop to inside the function which I'm not sure is any better haha 😅

# Issue 114 Updates
In the commit "added excel template function", I add a new function to the `WriteExcel` class called `_form_parameter_template(...)` which will create wide formatted template dataframes when writing out params that do not have data. A test for this is also added.  

## Questions 
2.  To create the wide formatted data, I needed to pass in the `input_data` to correctly index the columns as years. This involved changing the `WriteStrategy.write(...)` method to accept keyword arguments to allow for the passing of `input_data`. While I don't think this was a bad solution, Im not sure how clean it is either. I couldn't think of any other way to get the years for long formatted data to the `WriteExcel` class though? 

3. I slightly changed the logic of the the `_form_parameter(...)` method. I moved the check for empty dataframes out of this function and to where the function is called in the `WriteExcel.write_parameter(...)` method. While this means `_form_parameter(...)` will no longer catch if the input df is empty, since it is a private method and only used in this one place, I thought it improved readability. My question is, does this tradeoff make sense? Or is it better to keep those high level quick checks in the function?

trevorb1 commented on src/otoole/cli.py:
```diff
@@ -194,7 +196,7 @@ def conversion_matrix(args):
 
     if read_strategy and write_strategy:
         context = Context(read_strategy, write_strategy)
-        context.convert(args.from_path, args.to_path)
+        context.convert(args.from_path, args.to_path, input_data=input_data)
```
Example of question 2, showing the need to pass in `input_data` through to the write strategy.

trevorb1 commented on src/otoole/read_strategies.py:
```diff
@@ -125,6 +125,11 @@ def read(
 
             input_data[mod_name] = narrow
 
+        for config_type in ["param", "set"]:
+            input_data = self._get_missing_input_dataframes(
+                input_data, config_type=config_type
+            )
+
```
Example of question 1 where we loop over both `param` and `set`

trevorb1 commented on src/otoole/write_strategies.py:
```diff
@@ -81,11 +101,13 @@ def _write_parameter(
 
-        df = self._form_parameter(df, parameter_name, default)
         if not df.empty:
+            df = self._form_parameter(df, parameter_name, default)
             df.to_excel(handle, sheet_name=name, merge_cells=False, index=True)
         else:
-            logger.info(f"Skipped writing {parameter_name} as it is empty")
+            logger.debug(f"Dataframe {parameter_name} is empty")
+            df = self._form_parameter_template(parameter_name, **kwargs)
+            df.to_excel(handle, sheet_name=name, merge_cells=False, index=False)
```
Example of question 3, where I moved the check for for empty dataframes outside of the `_form_parameters(...)` method.
